enarx-keep-sev-lib/src/proxy.rs

```diff
@@ -5,120 +5,81 @@
+    pub num: Register<usize>,
 
-    /// Assign a value to the second parameter register (rsi).
-    pub fn second(mut self, val: Register<u64>) -> Self {
-        self.rsi = val;
-        self
-    }
-
-    /// Access the second parameter register (rsi).
-    pub fn get_second(&self) -> Register<u64> {
-        self.rsi
-    }
-
-    /// Assign a value to the third parameter register (rdx).
-    pub fn third(mut self, val: Register<u64>) -> Self {
-        self.rdx = val;
-        self
-    }
-
-    /// Access the third parameter register (rdx).
-    pub fn get_third(&self) -> Register<u64> {
-        self.rdx
-    }
-
-    /// Assign a value to the fourth parameter register (r10).
-    pub fn fourth(mut self, val: Register<u64>) -> Self {
-        self.r10 = val;
-        self
-    }
+    /// The syscall argument registers
+    ///
+    /// At most 6 syscall arguments can be provided.
+    pub arg: [Register<usize>; 6],
+}
 
-    /// Access the fourth parameter register (r10)
-    pub fn get_fourth(&self) -> Register<u64> {
-        self.r10
-    }
+/// A reply
+///
+/// The `Reply` struct is the most minimal representation of the register context
+/// needed for service replies from the hypervisor to the microkernel. An example
+/// of such a reply would be the return value from a proxied system call.
+///
+/// Although most architectures collapse this to a single register value
+/// with error numbers above `usize::max_value() - 4096`, `ppc64` uses
+/// the `cr0.SO` flag to indicate error instead. Unfortunately, we also
+/// can't use the built-in `Result` type for this, since its memory layout
+/// is undefined. Therefore, we use this layout with conversions for `Result`.
+#[repr(C)]
+#[derive(Copy, Clone, Default)]
+pub struct Reply {
+    val: Register<usize>,
+    err: bool,
+}
 
-    /// Assign a value to the fifth parameter register (r8).
-    pub fn fifth(mut self, val: Register<u64>) -> Self {
-        self.r8 = val;
-        self
+impl From<Result<Register<usize>, libc::c_int>> for Reply {
+    #[inline]
+    fn from(value: Result<Register<usize>, libc::c_int>) -> Self {
+        match value {
+            Ok(val) => Self { val, err: false },
+            Err(val) => Self {
+                val: Register::from_raw(val as usize),
+                err: true,
+            },
+        }
     }
+}
 
-    /// Access the fifth parameter register (r8).
-    pub fn get_fifth(&self) -> Register<u64> {
-        self.r8
+impl From<Reply> for Result<Register<usize>, libc::c_int> {
+    #[inline]
+    fn from(value: Reply) -> Self {
+        if value.err {
+            Err(value.val.raw() as _)
+        } else {
+            Ok(value.val)
+        }
     }
+}
 
-    /// Assign a value to the sixth parameter register (r9).
-    pub fn sixth(mut self, val: Register<u64>) -> Self {
-        self.r9 = val;
-        self
-    }
+/// A message, which is either a request or a reply
+#[derive(Copy, Clone)]
+pub union Message {
```

Comment:
Should this be `#[repr(C)]`? I'm guessing not because both of its variants are `#[repr(C)]`, right?

Reply:
This probably needs to be fixed. Otherwise the layout is undefined.